status/conditions.go

```diff
@@ -1,64 +1,68 @@
 package status
 
 import (
+	"context"
 	"time"
 
+	"github.com/awslabs/operatorpkg/event"
+	"github.com/awslabs/operatorpkg/object"
 	"github.com/prometheus/client_golang/prometheus"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-const (
-	metricLabelName            = "name"
-	metricLabelNamespace       = "namespace"
-	metricLabelKind            = "kind"
-	metricLabelConditionType   = "type"
-	metricLabelConditionStatus = "status"
-	metricNamespace            = "operator"
-	metricSubsystem            = "status_condition"
-)
-
-type ConditionedObject interface {
-	v1.Object
-	runtime.Object
-	GetConditions() Conditions
+func NewConditions(object Object, conditions []Condition) Conditions {
+	return Conditions{object: object, conditions: lo.ToPtr(conditions)}
 }
 
-type Conditions []v1.Condition
+type Conditions struct {
+	object     client.Object
+	conditions *[]Condition
+}
 
-func (c *Conditions) Set(conditionType string, conditionStatus v1.ConditionStatus, conditionMessage string) {
-	condition := c.Get(conditionType)
+func (c Conditions) Set(ctx context.Context, condition Condition) {
+	condition.ObservedGeneration = c.object.GetGeneration()
 
-	if condition == nil {
-		*c = append(*c, v1.Condition{
-			Type:               conditionType,
-			Status:             conditionStatus,
-			Message:            conditionMessage,
-			Reason:             conditionType,
-			LastTransitionTime: v1.Now(),
-		})
+	existing := c.Get(condition.Type)
+	if existing == nil {
+		condition.LastTransitionTime = metav1.Now()
+		*c.conditions = append(*c.conditions, condition)
 		return
 	}
-
-	if condition.Status != conditionStatus {
+	if condition.Status != existing.Status {
+		condition.LastTransitionTime = metav1.Now()
+		// Record a metric
 		ConditionDuration.
```

Comment:
What happens if the patch operation fails on updating this status condition? Will we fire the duration metric twice as well as event multiple times, which might skew the results

Reply:
Yes -- any thoughts on how to work around this?

Reply:
Discussed offline and changed from tracking the metric within the state transition itself to watching the state transition from a different controller.